distro.py

```diff
@@ -654,9 +654,10 @@ def __init__(self,
         self.root_dir = root_dir
         self.etc_dir = os.path.join(root_dir, 'etc') \
             if root_dir else _UNIXCONFDIR
-        self.os_release_file = os_release_file or \
-            os.path.join(self.etc_dir, _OS_RELEASE_BASENAME)
-        self.distro_release_file = distro_release_file or ''  # updated later
+        self.usr_lib_dir = os.path.join(root_dir, 'usr/lib') \
+            if root_dir else _UNIXLIBSDIR
+        self.os_release_file = os_release_file
```

Comment:
I worry about the change in semantics of `self.os_release_file` a bit. Previously, it was guaranteed to be the path to the file about to be tried to read. Now it can still be the empty string :thinking: 
I'm aware that constructors/initializers should not do much and that maybe doing the `os.path.isfile` detection loop (that I have seen further down) in here *could* not be ideal, I'm not sure. What do you think?

Reply:
I'm not sure about internal: It seems public API to me.

Before:
```console
# python -c 'import distro; print(distro.LinuxDistribution().os_release_file)'
/etc/os-release
```

After:
```console
# python -c 'import distro; print(distro.LinuxDistribution().os_release_file)'
<just a newline>
```

Reply:
What about something like this:

```python
if os_release_file:
    self.os_release_file = os_release_file
else:
    etc_dir_os_release_file = os.path.join(self.etc_dir, _OS_RELEASE_BASENAME) 
    usr_lib_os_release_file = os.path.join(self.usr_lib_dir, _OS_RELEASE_BASENAME)

    # NOTE: The idea is to respect order
    #       *and* have it set at all times for API backwards compatibility
    if os.path.isfile(etc_dir_os_release_file) or not os.path.isfile(usr_lib_os_release_file):
        self.os_release_file = etc_dir_os_release_file
    else:
        self.os_release_file = usr_lib_os_release_file
```

Reply:
Well, I haven't thought about this very attribute as being part of the API. IMHO, it should be private 🤷

Reply:
I'm not the one to decide. Without leading underscores I cannot say it's not public without get in conflict with myself. Any ideas how to proceed?

Reply:
Of course, it's currently _public_, but I guess it _should not_.
Two quick advanced GitHub search requests ([here](https://github.com/search?p=6&q=os_release_file+distro+language%3APython+NOT+filename%3Autil.py+NOT+filename%3Ametadata.py+NOT+filename%3Adistro.py+NOT+filename%3Autils.py&type=Code) & [there](https://github.com/search?q=distro.os_release_file+NOT+filename%3Autils.py&type=Code)) _seems_ to invalidate the fact that this attribute is _publicly_ being used (at least across GH's repositories).

Although, your idea to keep it up-to-date according to our internal logic is interesting too, we _only_ have to decide whether it's relevant or not.

Cheers 👋

Reply:
@nir0s we could benefit from your input here. Do you have a minute?

Reply:
@HorlogeSkynet I think in general our options are:

- (a) Treat `distro.LinuxDistribution.os_release_file` as current public API and change the PR to be more backwards compatible in semantics
- (b) Treat `distro.LinuxDistribution.os_release_file` as current public API by mistake, hide it away with leading underscores, increase the major version of distro to 2 to communicate the API break the semver way, and probably make a ton of other accidentally public API private while at it so that we don't break the API soon again.
- (c) Assume that `distro.LinuxDistribution.os_release_file` is of little interest to any user, ignore that it is current public API technically, change its current semantics, and see if we get (arguably justified) bug reports.

I have trouble standing behind (c) myself and (b) will probably not be fun or feasible, so I personally consider (a) the least worst.  Are there any other options?  What do you think?

Reply:
I think you're 100% right and I would not have come to these abstract leads by myself.
Of course `(a)` would be less _breaking_ than the others.
If I were a "proper" maintainer here, I'd have (cleaned up the code and) proposed an implementation for `(b)` I guess.
About `(c)`, maybe there are "others" here familiar with the code base/the project that would be able to argue in favor of it ?

Anyway, feel free to edit my branch as required, I don't mind.
Bye 👋

Reply:
I don't have editing rights outside my own pull requests here. If you'd be okay to go with (a) would making `__init__` do https://github.com/nir0s/distro/pull/262#discussion_r552161650 and then making `_os_release_info` do plain `return _open_and_return_info(self.os_release_file)` would be an option to you? That would be (a) in practice in my mind, or at least one way to do it. If we had state, I could add an approving review and then maybe it gets merged this month with some luck.  What do you think?

Reply:
Hey @hartwork, please find a new commit implementing `(a)` (and respecting `flake8` code-styles !). Bye, keep me posted :wave: